controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -242,6 +248,101 @@ private boolean greaterThan(Map<Long, Long> cut1, Map<StreamSegmentRecord, Integ
+                        ImmutableMap<StreamSegmentRecord, Integer> span2 = span2Future.join();
+                        return greaterThan(streamcut1, span1, streamcut2, span2);
+                    });
+    }
+
+    @Override
+    public CompletableFuture<StreamCutReferenceRecord> findStreamCutReferenceRecordBefore(Map<Long, Long> streamCut, RetentionSet retentionSet) {
+        Map<Set<Long>, ImmutableMap<StreamSegmentRecord, Integer>> fetched = new HashMap<>();
+        int size = retentionSet.getRetentionRecords().size();
+
+        if (retentionSet.getRetentionRecords().isEmpty()) {
+            return CompletableFuture.completedFuture(null);
+        }
+        return computeStreamCutSpan(streamCut)
+                .thenCompose(span1 -> {
+                    fetched.put(streamCut.keySet(), span1);
+                    BiFunction<StreamCutReferenceRecord, Boolean, CompletableFuture<Integer>> fn = (refRecord, comparison) -> {
+                        return getStreamCutRecord(refRecord)
+                                .thenCompose(record -> {
+                                    if (record.getStreamCut().equals(streamCut)) {
+                                        return CompletableFuture.completedFuture(0);
+                                    }
+                                    ImmutableMap<StreamSegmentRecord, Integer> sc = fetched.get(record.getStreamCut().keySet());
+                                    CompletableFuture<ImmutableMap<StreamSegmentRecord, Integer>> future;
+                                    if (sc != null) {
+                                        future = CompletableFuture.completedFuture(sc);
+                                    } else {
+                                        future = computeStreamCutSpan(record.getStreamCut())
+                                                .thenApply(span2 -> {
+                                                    fetched.put(record.getStreamCut().keySet(), span2);
+                                                    return span2;
+                                                });
+                                    }
+                                    return future.thenApply(span2 -> {
+                                        boolean compare = comparison ? greaterThan(streamCut, span1, record.getStreamCut(), span2) :
+                                                greaterThan(record.getStreamCut(), span2, streamCut, span1);
+                                        if (compare) {
+                                            return 1;
+                                        } else {
+                                            return -1;
+                                        }
+                                    });
+                                });
+                    };
+
+                    // binary search retention set. 
+                    return binarySearch(0, size, index -> {
```

Comment:
Instead of re-implementing `binarySearch` could we use Collections.binarySearch() ?

Reply:
this is an asynchronous binary search implementation. its working with futures :)